### PR TITLE
stats: dynamically update the average column size

### PR DESCRIPTION
**Pull request by fipped**
This PR is about the work of updating the tot_col_size of histograms when executing insert, delete and update statement. Therefore, a correct average column size can be computed when we use `show stats_histograms` statement.
The tot_col_size we dynamically maintain will be a little different from that computed by `analyze` statement since we get the size from the value directly instead of computing from the coded bytes, that is we leave out the length byte.
PTAL @lamxTyler @winoros

### CODE REVIEW

File: executor/write.go [hunk 1]
```diff
@@ -158,8 +158,11 @@ func updateRecord(ctx sessionctx.Context, h int64, oldData, newData []types.Datu
 	} else {
 		sc.AddAffectedRows(1)
 	}
-
-	ctx.GetSessionVars().TxnCtx.UpdateDeltaForTable(t.Meta().ID, 0, 1)
+	colSize := make(map[int64]int64)
+	for id, col := range t.WritableCols() {
```
Comment (lamxTyler) [inline executor/write.go]: This should also consider the column state?

File: executor/write.go [hunk 2]
```diff
@@ -158,8 +158,11 @@ func updateRecord(ctx sessionctx.Context, h int64, oldData, newData []types.Datu
...
+		colSize[col.ID] = int64(len(newData[id].GetBytes()) - len(oldData[id].GetBytes()))
```
Comment (lamxTyler) [inline executor/write.go]: No need to record the col size for fix length type.

File: statistics/handle_test.go [hunk 3]
```diff
@@ -335,16 +335,28 @@ func (s *testStatsCacheSuite) TestLoadHist(c *C) {
...
-		c.Assert(hist, Equals, newStatsTbl.Columns[id])
+		c.Assert(hist.TotColSize, Less, newStatsTbl.Columns[id].TotColSize)
+
+		temp := hist.TotColSize
+		hist.TotColSize = newStatsTbl.Columns[id].TotColSize
+		c.Assert(statistics.HistogramEqual(&hist.Histogram, &newStatsTbl.Columns[id].Histogram, false), IsTrue)
+		hist.TotColSize = temp
+
+		c.Assert(hist.CMSketch, NotNil)
+		c.Assert(newStatsTbl.Columns[id].CMSketch, NotNil)
```
Comment (lamxTyler) [inline statistics/handle_test.go]: It is already checked at `CMSketch.Equal`.

File: statistics/update.go [hunk 4]
```diff
@@ -190,6 +210,29 @@ func (h *Handle) dumpTableStatDeltaToKV(id int64, delta variable.TableDelta) (bo
...
+	for key, val := range delta.ColSize {
+		sql := fmt.Sprintf("update mysql.stats_histograms set version = %d, tot_col_size = tot_col_size + %d where hist_id = %d and table_id = %d", version, val, key, id)
+		_, err = h.ctx.(sqlexec.SQLExecutor).Execute(ctx, sql)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+	}
+	updated := h.ctx.GetSessionVars().StmtCtx.AffectedRows() > 0
+	_, err = h.ctx.(sqlexec.SQLExecutor).Execute(ctx, "commit")
+	return updated, errors.Trace(err)
```
Comment (lamxTyler) [inline statistics/update.go]: If the error is not nil, then updated will be always true.

File: statistics/update.go [hunk 5]
```diff
@@ -190,6 +210,29 @@ func (h *Handle) dumpTableStatDeltaToKV(id int64, delta variable.TableDelta) (bo
...
+	}
+	ctx := context.TODO()
+	_, err := h.ctx.(sqlexec.SQLExecutor).Execute(ctx, "begin")
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	version := h.ctx.Txn().StartTS()
+
+	for key, val := range delta.ColSize {
+		sql := fmt.Sprintf("update mysql.stats_histograms set version = %d, tot_col_size = tot_col_size + %d where hist_id = %d and table_id = %d", version, val, key, id)
```
Comment (lamxTyler) [inline statistics/update.go]: There may be two histograms with the same `hist_id` and `table_id`.

File: executor/write.go (hunk 2 above)
Comment (coocood) [inline executor/write.go]: `GetBytes` is only valid when the Kind of the Datum is KindString or KindBytes.

Maybe we need to add a method for Datum to return the size?

File: executor/write.go [hunk 6]
```diff
@@ -158,8 +158,13 @@ func updateRecord(ctx sessionctx.Context, h int64, oldData, newData []types.Datu
 	} else {
 		sc.AddAffectedRows(1)
 	}
-
-	ctx.GetSessionVars().TxnCtx.UpdateDeltaForTable(t.Meta().ID, 0, 1)
+	colSize := make(map[int64]int64)
+	for id, col := range t.WritableCols() {
+		if col.State == model.StatePublic && newData[id].Size() != 0 {
+			colSize[col.ID] = int64(newData[id].Size() - oldData[id].Size())
```
Comment (lamxTyler) [inline executor/write.go]: This is not right? If the new data happens to be null, but the old data is not null.

File: executor/write.go [hunk 7]
```diff
@@ -385,7 +390,13 @@ func (e *DeleteExec) removeRow(ctx sessionctx.Context, t table.Table, h int64, d
 	}
 	ctx.StmtAddDirtyTableOP(DirtyTableDeleteRow, t.Meta().ID, h, nil)
 	ctx.GetSessionVars().StmtCtx.AddAffectedRows(1)
-	ctx.GetSessionVars().TxnCtx.UpdateDeltaForTable(t.Meta().ID, -1, 1)
+	colSize := make(map[int64]int64)
+	for id, col := range t.WritableCols() {
+		if col.State == model.StatePublic {
+			colSize[col.ID] = -int64(data[id].Size())
```
Comment (lamxTyler) [inline executor/write.go]: If the `Size` is 0, it is not needed to update it?

File: table/tables/tables.go [hunk 8]
```diff
@@ -434,7 +434,13 @@ func (t *Table) AddRecord(ctx sessionctx.Context, r []types.Datum, skipHandleChe
 		}
 	}
 	sessVars.StmtCtx.AddAffectedRows(1)
-	sessVars.TxnCtx.UpdateDeltaForTable(t.ID, 1, 1)
+	colSize := make(map[int64]int64)
+	for id, col := range t.WritableCols() {
+		if col.State == model.StatePublic {
+			colSize[col.ID] = int64(r[id].Size())
```
Comment (lamxTyler) [inline table/tables/tables.go]: Ditto.

File: types/datum.go [hunk 9]
```diff
@@ -122,6 +122,11 @@ func (d *Datum) IsNull() bool {
 	return d.k == KindNull
 }
 
+// Size gets the length of byte of the datum.
+func (d *Datum) Size() int {
+	return len(d.b)
```
Comment (coocood) [inline types/datum.go]: We need to switch on `d.k` to determine the size of the datum.

Comment (coocood) [reply types/datum.go]: Or if we only care about varlen size, then we don't need this method, use `len(d.GetBytes())` will do.

**Comment by shenli** [comment]
@fipped Please resolve the conflicts.

**Comment by shenli** [comment]
@lamxTyler PTAL

File: statistics/update.go [hunk 10]
```diff
@@ -154,10 +162,19 @@ func (h *Handle) DumpStatsDeltaToKV() error {
...
 		if err != nil {
 			return errors.Trace(err)
 		}
+		if updated {
+			h.globalMap.update(id, -item.Delta, -item.Count, nil)
+		}
+		if err = h.dumpTableStatColSizeToKV(id, item); err != nil {
+			return errors.Trace(err)
+		}
+		m := h.globalMap[id]
```
Comment (lamxTyler) [inline statistics/update.go]: We do not need to update the map if the `updated` is true, just delete it.

**Comment by fipped** [comment]
PTAL @coocood

File: sessionctx/variable/session.go [hunk 11]
```diff
@@ -101,13 +101,21 @@ type TransactionContext struct {
 }
 
 // UpdateDeltaForTable updates the delta info for some table.
-func (tc *TransactionContext) UpdateDeltaForTable(tableID int64, delta int64, count int64) {
+func (tc *TransactionContext) UpdateDeltaForTable(tableID int64, delta int64, count int64, colSize *map[int64]int64) {
```
Comment (coocood) [inline sessionctx/variable/session.go]: We don't need to use pointer for a map.

File: executor/write.go [hunk 12]
```diff
@@ -165,8 +165,16 @@ func updateRecord(ctx sessionctx.Context, h int64, oldData, newData []types.Datu
 	} else {
 		sc.AddAffectedRows(1)
 	}
-
-	ctx.GetSessionVars().TxnCtx.UpdateDeltaForTable(t.Meta().ID, 0, 1)
+	colSize := make(map[int64]int64)
+	for id, col := range t.WritableCols() {
```
Comment (coocood) [inline executor/write.go]: We can just range over `t.Cols()`, it will return public columns only.

File: executor/write.go [hunk 13]
```diff
@@ -392,7 +400,16 @@ func (e *DeleteExec) removeRow(ctx sessionctx.Context, t table.Table, h int64, d
 	}
 	ctx.StmtAddDirtyTableOP(DirtyTableDeleteRow, t.Meta().ID, h, nil)
 	ctx.GetSessionVars().StmtCtx.AddAffectedRows(1)
-	ctx.GetSessionVars().TxnCtx.UpdateDeltaForTable(t.Meta().ID, -1, 1)
+	colSize := make(map[int64]int64)
+	for id, col := range t.WritableCols() {
```
Comment (coocood) [inline executor/write.go]: ditto

**Comment by coocood** [comment]
LGTM

**Pull request merged**